### ISSUE TITLE
apply to plexon data

### ISSUE THREAD

**Issue by happyqiu** [csn-le/wave_clus#161]
Hi ferchaure,

Sorry to disturb you, I'm new to the spike sorting, and try to apply your method to the plexon data, but got some, maybe quite simple questions. I extracted the waveforms('spikes') and timestamps('index') from .plx file and saved as .mat file. But each waveform only contains 56 sample points. When I run the code 'Get_spikes.m', I was always told:
Using detected spikes Segment read only for plotting
0 of 1 'spikes' files done.

I've changed:
par.w_pre = 8, par.w_post = 24. Are these parameters work for the algorithm?

As I understand, Get_spikes.m will help us to find the peak and align the waveforms, right? Do I need to write another code to align them manually?

Thanks a lot!

**Comment by ferchaure** [comment]
> par.w_pre = 8, par.w_post = 24. Are these parameters work for the algorithm?

yes, I guess that you have a sampling rate of 15 kHz or less?

> Get_spikes.m will help us to find the peak and align the waveforms, right?

Yes, check the polarity of the spikes and the par.detection parameter.

**Comment by happyqiu** [comment]
Thank you so much for your reply. The frequency for spike recording is 40000Hz. Here I attached an example of my *_spike.mat file. 
The number of pre-threshold points for each waveform is 12. And I have set the par.detection = 'neg'. As for the detection parameters, I mainly need to change the number of points and detection, rights?

[20_spikes.zip](https://github.com/csn-le/wave_clus/files/4299469/20_spikes.zip)
![Screenshot (2)](https://user-images.githubusercontent.com/50755486/76106844-8d5d8d00-5f8c-11ea-99f8-ff3b45023a6f.png)

**Comment by ferchaure** [comment]
The spikes look a bit odd but the parameters are ok.

**Comment by happyqiu** [comment]
Hi ferchaure,

I'm sorry, but I'm still kind of confused about this package. 
1) Will the template-matching be processed only if I click the 'Force' button? 
2) I set the parameter par.max_spk = 30000; Does it mean that only 30000 spikes will be randomly chosen and clustered unless I click the 'Force' button to do the rest?
3) What do the different template_types stand for, like 'center','mahal'...?
4) In my case, the spike waveforms are already generated based on the threshold detection, and they're not aligned by the 'neg' peaks, because the neurons we're interested in may have different latencies to certain stimuli. Does it matter when using your algorithm?

Thanks,
Yuechen

**Comment by ferchaure** [comment]
1. Yes
2. Yes, or even less. In this method, the cluster is used to find a clean mean template, some spikes are just too noisy and the clustering won't assign them. 
3. 'center' is the more obscure (and for retro-compatibility I didn't change it). Is a simple Euclidean distance between the template and the spikes. Check the reference paper for a description of the maximum distance. 'mahal' is the same but using the mahalanobis distance.
4. Yes, because the distances are susceptible to the alignment. I would rename the spikes files to something without '_spikes.mat' at the end and then use the Get_spikes() function to re-align the peaks and make a new '_spikes.mat' file with aligned spikes.

**Comment by happyqiu** [comment]
Thanks a lot for your reply! But as for the last question, won't they be separated better if these two kinds of waveforms in the picture are not aligned? I think the latency information will also be considered as some kind of feature for clustering?

![wf](https://user-images.githubusercontent.com/50755486/76789231-a8dd4a80-6779-11ea-9a6a-4f5fb560a607.jpg)

**Comment by ferchaure** [comment]
> won't they be separated better if these two kinds of waveforms in the picture are not aligned? 

No, because the important part is that all the spikes from a neuron have a small euclidean distance. If you don't align, the spikes from the same neuron could have a great distance.

> I think the latency information will also be considered as some kind of feature for clustering?

This latency information as you defined it, it's quite noisy. A small amount of noise could generate a difference of a few samples and increase a lot the variability of the clusters. Or even worst, the clustering could detect multiple clusters for the same neuron but with a different number of samples shifted.

I understand your idea, this "latency information" could be useful if you can guarantee that the sample of threshold crossing is exactly the same for all the spikes of a neuron, but that is never possible.

**Comment by happyqiu** [comment]
Hi Ferchaure,

Sorry to trouble you again. Currently I'm trying to sort out as many clusters as the algorithm can, because I found that many spikes included in the cluster 0 might also be real unit spikes. 
I tried changing several parameters to make it, like template_sdnum (max radius of cluster in std devs), min_clus(minimum size of a cluster). And got a few more questions. 

1. If template_sdnum is smaller, there should be more spikes that be assigned to specific clusters, right?
2. I also want to try 'pca' method, but found in the 'force' process, there was no 'pca' option. Is it okay to use pca to get the template and then apply wavelet to force additional membership?

3. Considering that ending points of the spike waveforms may affect too much on the sorting process, I try to use w_pre = 8 and w_post = 24 to align the spikes (originally there are 56 points in each waveform), but got the following error:
   Unable to perform assignment because the size of the left side is 155375-by-12 and the
   size of the right side is 155375-by-10.

Error in spike_alignment (line 30)
    spikes1(:,1:align_window+2) = -spikes(:,align_window:-1:1);

It seems that this line won't be right at anytime. I think it should be
line 30    spikes1(:,1:align_window) = -spikes(:,align_window:-1:1);
line 31    spikes1(:,1+align_window:(2*align_window+ls)) = spikes(:,1:align_window+ls); ?

Also, could you provide any other suggestions about how to get more clusters if possible?
Thanks a lot!

**Comment by ferchaure** [comment]
> If template_sdnum is smaller, there should be more spikes that be assigned to specific clusters, right?

No, it will be fewer spikes assigned to specific clusters (more on the class 0)

> I` also want to try 'pca' method, but found in the 'force' process, there was no 'pca' option. Is it okay to use pca to get the template and then apply wavelet to force additional membership?

You can use any method to get the features (I recommend the wavelets). But Wave_clus always uses the whole waveform (not features) to force additional membership.

About the error, our code is fine (you need extra samples to make a proper interpolation later). When you define the w_pre and w_post parameters, take notice that from the waveform of the detected spikes: 2 samples from the beginning and other 2 from the end will be "removed".  After that, the waveform should have w_pre samples before the peak (including the peak) and w_post samples after the peak. If the spikes don't have enough samples some error will rise.

**Comment by happyqiu** [comment]
Hi ferchaure, I'm sorry to disturb you again. Is there a threshold included in the algorithm to distinguish multi-unit and single neuron activity? It seems cluster 1 isn't always the multi-unit activity,right?

**Comment by ferchaure** [comment]
Hi, We don't have anything like that, because using a multi-unit flag could be quite tricky sometimes, especially if it is only a threshold. You are right, the cluster 1 is usually a multi-unit but just that.

An automatic option could use cluster quality metrics to label clusters (there are a few papers about that topic in spike sorting).

**Comment by happyqiu** [comment]
Thanks for your reply ! But in my attached case, it looks obvious that cluster 1 isn't multi-unit. So we actually don't know and should check manually, right?
![fig2print_wf_sig20](https://user-images.githubusercontent.com/50755486/80314354-f64fcd00-87a5-11ea-80f5-22ca7744d5e9.png)

**Comment by ferchaure** [comment]
Yes, you should check, but I would say that cluster 2 is the multi-unit: Low amplitude and a waveform that is just a bump crossing the detection threshold. 

Are you using the old version of the method? Probably the new version of wave_clus could get cleaner results.

**Comment by happyqiu** [comment]
Thanks a lot for your quick response. I think I downloaded the latest version from the Github. Or would  you mind sending me the link of the new version if it's convenient? In the results above, maybe it's because I set the minimum size of a cluster as 60, then only 3 clusters are found?

**Comment by ferchaure** [comment]
Do You set min_clus = 60?  that is too hight (that is why I presumed that it was the older version ). That number is not only the minimum theoretical size, but it is also the size of the minimum jump in the temperature plot to explore that solution.
